### ISSUE TITLE
A Bug In rate-limiting

### ISSUE THREAD

**Issue by wangdafang** [Kong/kong#6810]
My question:rate-limiting is inaccurate in slowly upstream interface (which have long response time)

My Kong version is 1.4.3(and rate-limiting code is the same with lasted code)
rate-limiting config :
second 1
minute nil
hour nil

I have two scenarios for test:
1.request 2 counts in 1 second (upstream target response in 1ms.)
result:
first http-response is 200
second http-response is 429

2.request 2 counts in 1 second (upstream target response in 1s.)
result:
first http-response is 200
second http-response is 200

From the result,we can see,In second scenario,the result is unexpected.
And I have found the rate-limiting's code,the question is happend on code
'
kong.ctx.plugin.timer = function()
      local ok, err = timer_at(0, increment, conf, limits, identifier, current_timestamp, 1)
      if not ok then
        kong.log.err("failed to create timer: ", err)
      end
    end
'
We all no that ngx.timer.at will run after this thread is finish.But it's not suitable for rate-limiting.
I need for your help

**Comment by ttttthub** [comment]
It  has no relation with upstream target response time.

**Comment by wangdafang** [comment]
But I have found,if upstream response time is very long,the func'increment' will not exec before upstream response finish

**Comment by ttttthub** [comment]
When a timer expires, the user Lua code in the timer callback is running in a "light thread" detached completely from the original request creating the timer. So objects with the same lifetime as the request creating them, like cosockets, cannot be shared between the original request and the timer user callback function.

**Comment by wangdafang** [comment]
I know what you mean.But I got the result can't support your viewpoint.
Let's read my test case:
--------------------------------------------------------------------
2021/02/07 14:10:12 [error] 23404#0: *95873 [kong] handler.lua:84 [rate-limiting] rate-limiting:id:904, client: 10.0.1.27, server: www.test.com, request: "GET /jedis/get/byte HTTP/1.1", host: "www.test.com"
2021/02/07 14:10:13 [error] 23407#0: *95944 [kong] handler.lua:84 [rate-limiting] rate-limiting:id:904, client: 10.0.2.27, server: www.test, request: "GET /jedis/get/byte HTTP/1.1", host: "www.test.com"
2021/02/07 14:10:23 [error] 23404#0: *97353 [kong] handler.lua:73 rate-limiting add count finish, context: ngx.timer, client: 10.0.2.27, server: 0.0.0.0:80
2021/02/07 14:10:23 [error] 23407#0: *97354 [kong] handler.lua:73 rate-limiting add count finish, context: ngx.timer, client: 10.0.2.27, server: 0.0.0.0:80
--------------------------------------------------------------------
"rate-limit add count finish" is write in func increment's first line
"[rate-limiting] rate-limiting:id:904" is write in func RateLimitingHandler:access's first line

rate-limiting:
config:
policy local
second 1
minute nil
hour nil
I used "curl http://www.test.com/jedis/get/byte" in item, and exec twice in one minute.
And I make the upstream server block until I unblock it after some second.

As you see , in the log,when I unblock the upstream server, the "rate-limit add count finish" is apper in my log.And this happens every time
